src/core/thread/mle_router.cpp

```diff
@@ -1371,6 +1383,34 @@ ThreadError MleRouter::HandleAdvertisement(const Message &aMessage, const Ip6::M
         ExitNow();
 
     case kDeviceStateRouter:
+        // check current active router number
+        routerCount = 0;
+
+        for (uint8_t i = 0; i < kMaxRouterId; i++)
+        {
+            if (route.IsRouterIdSet(i))
+            {
+                routerCount++;
+            }
+        }
+
+        if (routerCount > mRouterDowngradeThreshold)
+        {
+            if (HasMinDowngradeNeighborRouters() &&
+                HasOneNeighborwithComparableConnectivity(route, routerId) &&
+                HasSmallNumberOfChildren())
+            {
+                if (mRouterSelectionJitterTimeout == 0)
+                {
+                    mRouterSelectionJitterTimeout = otPlatRandomGet() % kRouterSelectionJitter;
+                    mStateUpdateTimer.Start(kStateUpdatePeriod);
+                    ExitNow();
```

Comment:
Not sure exit is correct.  We still want to call `UpdateRoutes()` and `SendServerDataNotification()` below.